### PR TITLE
Support parallel function calling in pydantic program

### PR DESCRIPTION
**Pull request by Disiok**
# Description

Support parallel function calling in pydantic program

### CODE REVIEW

File: llama_index/program/openai_program.py [hunk 1]
```diff
@@ -33,6 +40,39 @@ def _get_json_str(raw_str: str, start_idx: int) -> Tuple[Optional[str], int]:
...
+) -> Union[BaseModel, List[BaseModel]]:
+    outputs = []
+    for tool_call in tool_calls:
+        function_call = tool_call["function"]
+        if verbose:
+            name = function_call["name"]
+            arguments_str = function_call["arguments"]
+            print(f"Function call: {name} with args: {arguments_str}")
+
+        if isinstance(function_call["arguments"], dict):
```
Comment (jerryjliu) [inline llama_index/program/openai_program.py:57]: ooc when would it be one or the other

**Review by jerryjliu — APPROVED** [review]

File: llama_index/program/openai_program.py (hunk 1 above)
Comment (Disiok) [reply llama_index/program/openai_program.py:57]: this is existing code we have. I think it was due to difference in OpenAI and AzureOpenAI behavior in the past.

**Review by nerdai — APPROVED** [review]
lgtm!

Potentially nit/silly question: but is there a reason we need both `Model` and `BaseModel`?

**Comment by Disiok** [comment]
> lgtm!
> 
> Potentially nit/silly question: but is there a reason we need both `Model` and `BaseModel`?

Took a look again, the `Model` is a `TypeVar`, but looks like we are not using it properly to type the outputs. Thanks for pointing this out, will fix.

**Pull request merged**